File: medium/trailfact.cpp
#include <iostream>
#include <cstdint>

class Solution
{
public:
    int trailingZeroes(int n)
    {
        int res = 0;
        std::uint64_t f = 1;
        for (int i = 1; i <= n; i++)
        {

            int j = i;
            while (j % 10 == 0 && j >= 10)
            {
                std::cout << "Before mult " << i << '\n';
                j /= 10;
                res++;
            }

            f *= j;

            std::cout << f << '\n';

            while (f % 10 == 0)
            {
                std::cout << "After mult " << i << '\n';
                res++;
                f /= 10;
            }
        }

        return res;
    }
};

int main()
{
    std::cout << Solution().trailingZeroes(30) << '\n';
}

File: medium/trailfact.py

class Solution:
    def trailingZeroes(self, n: int) -> int:
        
        res = 0
        f = 1
        twos = 0
        fives = 0
        for i in range(1, n + 1):
            
            j = i
            while (j >= 5 and j % 5 == 0):
                fives += 1
                j //= 5
            
            while (j >= 2 and j % 2 == 0):
                twos += 1
                j //= 2

            f *= j

            while (twos >= 1 and fives >= 1):
                res += 1
                twos -= 1
                fives -= 1
        
        return res
    
print( Solution().trailingZeroes(30) )
